File: Sorting Algorithms/Quick Sort/QuickSort.py
def main(): 
    print("Quick Sort Algorithm using python")

    arr: list[int] = [2, 1, 5, 8, 3, 9, 4, 7, 6, 0]

    print("\nArray before sorting: ",arr)
    quick_sort(arr, 0, arr.__len__()-1)
    print("\nArray after sorting: ",arr)


def quick_sort(arr: list[int], start: int, end: int) -> None:
    """Sorts a list of integers by finding correct position of next int and swapping."""
    if start < end:
        pivot = partition(arr, start, end)
        quick_sort(arr, start, pivot-1)
        quick_sort(arr, pivot+1, end)
    return 

def partition(arr: list[int], start: int, end: int) -> int:
    x: int = arr[end]
    i: int = start - 1

    for j in range(start,end):
        if arr[j] <= x:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]

    arr[i+1], arr[end] = arr[end], arr[i+1]

    return i+1


if __name__ == "__main__":
    main()

File: Sorting Algorithms/Quick Sort/QuickSort.java
class QuickSort{


    public static void main(String[] args){
        int arr[] = {2, 1, 5, 8, 3, 9, 4, 7, 6, 0};
        int n = arr.length;

        System.out.println("Quick Sort Algorithm using Java");

        System.out.print("\n\nArray before sorting: [ ");
        for(int i=0; i<n-1; i++){
            System.out.print(arr[i] + ", ");
        }
        System.out.println(arr[n-1]+ " ]");

        quickSort(arr, 0, n-1);

        System.out.print("\n\nArray after sorting: [ ");
        for(int i=0; i<n-1; i++){
            System.out.print(arr[i] + ", ");
        }
        System.out.println(arr[n-1]+ " ]");
    }


    public static void quickSort(int[] arr, int start, int end){
        int pivot;
        if(start < end){
            pivot = partition(arr, start, end);
            quickSort(arr, start, pivot-1);
            quickSort(arr, pivot+1, end);
        }
    }

    public static int partition(int[] arr, int start, int end){
        int x = arr[end];
        int i = start-1;
        int temp;
        
        for(int j = start; j < end; j++){
            if(arr[j] <= x){
                i++;
                temp = arr[j];
                arr[j] = arr[i];
                arr[i] = temp;
            }
        }
        temp = arr[i+1];
        arr[i+1] = arr[end];
        arr[end] = temp;

        return i+1;
    }
    

}